feat: Add ability to hijack-cursor

ghostbuster91 opened the pull request:
This closes #604 

TODO:
- [x] docs
- [x] config
- [x] proper initialization

ghostbuster91 commented on lua/neo-tree/setup/init.lua:
```diff
@@ -728,6 +728,8 @@ M.merge_config = function(user_config, is_auto_config)
   local rt = utils.get_value(M.config, "resize_timer_interval", 50, true)
   require("neo-tree.ui.renderer").resize_timer_interval = rt
+  --TODO how to initialize properly
+  require("neo-tree.sources.common.hijack_cursor").setup(M.config, nil)
```
I wasn't able to figure out how this should be initialized. I would be grateful for some guidance here.

cseickel replied:
Generally all initialization that is needed happens in setup.merge_config. There's no particular pattern to that except for the sources themselves that have to conform to an api.

ghostbuster91 commented on lua/neo-tree/sources/common/hijack_cursor.lua:
```diff
@@ -0,0 +1,38 @@
+---@param config table Configuration table containing any keys that the user
+---wants to change from the defaults. May be empty to accept default values.
+M.setup = function(config, global_config)
+  local modules = { "filesystem", "git_status", "buffers" }
+  for _, module in ipairs(modules) do
```
Is this the proper way to initialize something for all sources?

cseickel replied:
No, because the list of sources is not a static thing and will change according to what extensions a user has. The best place to do it right now would be in this block:

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/8c89efb93b8383666a6f989762c2f610f5f2c930/lua/neo-tree/setup/init.lua#L671-L682

ghostbuster91 replied:
Done :+1:

ghostbuster91 commented:
@cseickel Thanks for the quick review. I have addressed your comments. Is there anything left that I should do? Initially I was thinking also about mentioning it in the vim docs, but I see that not all of the flags are there, so I am not sure if I should put it there.

cseickel commented on lua/neo-tree/sources/common/hijack_cursor.lua:
```diff
@@ -0,0 +1,36 @@
+
+local M = {}
+
+local setup_for_module = function(module)
```
I feel like this function name can be named better. Maybe something like `create_handler_for_module`?

ghostbuster91 replied:
done

cseickel commented on lua/neo-tree/sources/common/hijack_cursor.lua:
```diff
@@ -0,0 +1,36 @@
+
+local setup_for_module = function(module)
+  return function()
+    local state = manager.get_state(module)
```
This will actually only give you the state of the sidebar for the current tab. There may also be state objects for other tabs along with many state objects for windows where position = "current".

There is a `manager._for_each_state()` function that you can use for this:
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/8c89efb93b8383666a6f989762c2f610f5f2c930/lua/neo-tree/sources/manager.lua#L54

ghostbuster91 replied:
Nice, this simplifies everything :)

ghostbuster91 replied:
Although, this doesn't seem to work for some reason. Any idea what could that be?

ghostbuster91 replied:
It seems that `all_states` is not populated in the `manager` at the time I initialize hijack.

cseickel replied:
I'm sorry, I kind of mislead you here. This doesn't really work because states are created and destroyed as the user opens new trees in new tabs/windows and destroys those tabs/windows.

The event should just be subscribed once, and then every time it fires you should figure out which neo-tree window you are in (if any), then retrieve the state for that window or tab and move the cursor.

You can tell if a neo-tree window is focused by checking the buffer local variables. This code is an example of getting state for the current window:
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/8c89efb93b8383666a6f989762c2f610f5f2c930/lua/neo-tree/setup/init.lua#L282-L287

If the position IS current, the get_state call needs to include the window id.

cseickel replied:
Here's one example of subscribing to an event, there are actually a lot of subscriptions in the setup/init methods:

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/7f6fa04dbd8e8c79d1af33bc90e856b65d8641da/lua/neo-tree/setup/init.lua#L704-L708

You can also view the help file for more details about the events and arguments:
```
:h neo-tree-events
```

ghostbuster91 replied:
all clear now, thanks

cseickel reviewed (commented):
As far as docs go, not everything is in the help file because I was just too lazy to write docs for every single option. The general rule is that anything that can easily be explained with an inline comment in the config doesn't need to be explicitly listed in the help file.

If you wanted to put in a doc entry, you could just start a new section that could someday contain all of the options that just need one-line explanations like this one. Or you can skip it.

ghostbuster91 commented:
Hi @cseickel, it seems that I can now get back to this one, though I still don't know where to subscribe for the events. Could you drop some hint? :pray:

cseickel commented on lua/neo-tree/sources/common/hijack_cursor.lua:
```diff
@@ -0,0 +1,38 @@
+      log.debug("Cursor moved in tree window, hijacking cursor position")
+      local cursor = vim.api.nvim_win_get_cursor(0)
+      local row = cursor[1]
+      local current_line = vim.api.nvim_get_current_line()
```
I think this could be:
```suggestion
      local current_line = cursor[0]
```

ghostbuster91 replied:
I think that you meant `cursor[2]` but that returns only the column position. Here I need to fetch the text for the current line.

cseickel replied:
You're right, I was very confused. I was thinking that you were looking for the current line number for some reason, and also that tables were zero indexed.

cseickel reviewed (commented):
I'm not sure why, but it does not seem to work with `position = "current"`.

ghostbuster91 commented on lua/neo-tree/sources/common/hijack_cursor.lua:
```diff
@@ -0,0 +1,38 @@
+    if vim.o.filetype ~= "neo-tree" then
+        return
+    end
+    local source = vim.api.nvim_buf_get_var(0, "neo_tree_source")
```
Good idea, done.

ghostbuster91 commented:
> I'm not sure why, but it does not seem to work with `position = "current"`.

I will look into that.

ghostbuster91 commented:
@cseickel I looked into that. It doesn't work because `state.winid` is `nil`. I thought that maybe this is some special case that needs to be handled. 
I found this https://github.com/ghostbuster91/neo-tree.nvim/blob/11f6042509d8e4dc3ffbddf0bf1d8b7d01ac3d94/lua/neo-tree/ui/selector.lua#L254 and made a similar change. However after that change neotree failed because `state.tree` is also `nil` :eyes: (in this line https://github.com/ghostbuster91/neo-tree.nvim/blob/11f6042509d8e4dc3ffbddf0bf1d8b7d01ac3d94/lua/neo-tree/sources/common/hijack_cursor.lua#L21) 
Any idea what is happening here?

cseickel commented:
> Any idea what is happening here?

Yes. The problem is that if you call `manager.get_state()` without specifying a window id, you will get back the sidebar state for the tab. What you need to do is first check if the window was open in `position = "current"` or not, and if the position is `"current"` then you need to pass the winid as the third parameter to `get_state(source, tabid, winid)` to get the state for that window and not the tab.

You'd need something like this:

```lua
local winid = nil
local _, position = pcall(vim.api.nvim_buf_get_var, 0, "neo_tree_position")
if position == "current" then
  winid = vim.api.get_current_win()
end
local state = manager.get_state(source, nil, winid)
```

That block of logic should probably be a utility function, or mayabe the `get_state` function itself should be checking if the currently focused window is a neo-tree window and doing that work for you.

ghostbuster91 commented:
I see, thanks for the explanation. For now I will just add this check here but I think that it would make sense to move it eventually to `get_state`. I can do that as a follow-up.

cseickel commented:
Hey @ghostbuster91, sorry I think I dropped the ball on this. Is this good to go?

ghostbuster91 commented:
@cseickel I fixed the issue with `position=current`. Could you give it a try?

cseickel reviewed (approved):
Looks good and it works for me in all situations. Thanks!
